### PR TITLE
Quota Rate Limiting

### PR DESCRIPTION
**Pull request by AnatoleAM**
Creating a rate limiter for GQL which uses a Quota system rather than a conventional request-based rate limit.

Each request is granted a quota, with a specific number of points. Each queried field decreases the points by a set amount. For calls to the same object, the point cost is less, while relational calls will cost more. Ultimately, if the amount of points remaining reaches zero, the request is rejected with status `429 Too Many Requests` and information about the quota.

This adds the following headers to requests:
- `X-Quota-Limit` is the initial amount of points in the Quota
- `X-Quota-Remaining` is the amount of points left over after the request (may be negative if exceeded)
- `X-Quota-Usage` offers details about each queried fields and their quota costs as a JSON map.

### CODE REVIEW

File: src/configure/config.go [hunk 1]
```diff
@@ -32,6 +32,8 @@ type Config struct {
 		URI              string `mapstructure:"uri" json:"uri"`
 		Type             string `mapstructure:"type" json:"type"`
 		OauthRedirectURI string `mapstructure:"oauth_redirect_uri" json:"oauth_redirect_uri"`
+
```
Comment (TroyDota) [inline src/configure/config.go:35]: Extra newline?

File: src/server/v3/gql-v3.go [hunk 2]
```diff
@@ -32,49 +41,65 @@ func (*Query) HelloWorld() string {
...
+		if badQueries, _ := gCtx.Inst().Redis.RawClient().HLen(ctx, fmt.Sprintf("%s:blocked-queries:client-ip:%s", instance.RedisPrefix, clientIP)).Result(); badQueries >= 25 {
+			return c.Status(fiber.StatusForbidden).JSON(&fiber.Map{
+				"status": fiber.ErrForbidden,
+				"error":  "You are temporarily blocked from using this API",
+			})
+		}
+
+		// Check if the query is allowed
+		h := sha256.New()
+		h.Write(utils.S2B(strings.Map(func(r rune) rune {
```
Comment (TroyDota) [inline src/server/v3/gql-v3.go:83]: What does this do?

File: src/server/v3/helpers/errors.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package errors
+package helpers
```
Comment (TroyDota) [inline src/server/v3/helpers/errors.go:1]: What does this file do?

File: src/server/v3/helpers/keys.go [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package helpers
```
Comment (TroyDota) [inline src/server/v3/helpers/keys.go:1]: You actually dont need these.

File: src/server/v3/helpers/quota.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	C    *fiber.Ctx
+	// The initial amount of Quota Points
+	Limit int `json:"limit"`
+	// Quota Points available
+	Points int `json:"points"`
+	// Map of fields to quota costs
+	Fields map[string]int
+}
+
+func (q *Quota) DecreaseByOne(object string, field string) bool {
```
Comment (TroyDota) [inline src/server/v3/helpers/quota.go:21]: You actually can do this via atomic, without using a mxtx, you can do where limit and Points are both pointers and then use atomic.AddInt32 or smth

File: src/server/v3/gql-v3.go (hunk 2 above)
Comment (AnatoleAM) [reply src/server/v3/gql-v3.go:84]: Gonna leave this for now until we have a better way of building the hash

File: src/server/v3/resolvers/mutation/emote.go [hunk 6]
```diff
@@ -18,7 +18,7 @@ func (r *Resolver) Emote(ctx context.Context, args struct {
 	// Get current user
 	user := ctx.Value("user").(*structures.User)
```
Comment (TroyDota) [inline src/server/v3/resolvers/mutation/emote.go:19]: monkaS ?

File: src/server/v3/gql-v3.go [hunk 7]
```diff
@@ -32,49 +42,64 @@ func (*Query) HelloWorld() string {
...
-		ctx := context.WithValue(context.Background(), utils.Key("user"), c.Locals("user")) // Add auth user to context
-		ctx = context.WithValue(ctx, utils.Key("request"), c)                               // Add request to context
+		defaultQuota := gCtx.Config().Http.QuotaDefaultLimit
+		quota := &helpers.Quota{
+			C:      c,
+			Limit:  &defaultQuota,
+			Points: &defaultQuota,
+			Fields: sync.Map{},
+		}
+		ctx := context.WithValue(context.Background(), helpers.UserKey, c.Locals("user")) // Add auth user to context
```
Comment (TroyDota) [inline src/server/v3/gql-v3.go:68]: remove helpers and rather just use `utils.Key("user")`

File: src/server/v3/gql-v3.go [hunk 8]
```diff
@@ -98,6 +123,35 @@ func GQL(gCtx global.Context, app fiber.Router) {
...
+
+			c.Set("X-Quota-Usage", utils.B2S(b))
+		}
+		if !quota.Check() {
+			// Temporarily block this query
+			pipeline := gCtx.Inst().Redis.RawClient().Pipeline()
+			pipeline.SetEX(ctx, redisKey, "", time.Hour)
+			pipeline.HSet(ctx, fmt.Sprintf("%s:blocked-queries", instance.RedisPrefix), qh, clientIP)
+			pipeline.HSet(ctx, fmt.Sprintf("%s:blocked-queries:client-ip:%s", instance.RedisPrefix, clientIP), qh, req.Query)
+			pipeline.Exec(ctx)
```
Comment (TroyDota) [inline src/server/v3/gql-v3.go:146]: check errors?

File: src/server/v3/helpers/keys.go (hunk 4 above)
Comment (TroyDota) [reply src/server/v3/helpers/keys.go:1]: Remove this file.

File: src/server/v3/resolvers/query/emote.go [hunk 9]
```diff
@@ -32,6 +34,7 @@ func CreateEmoteResolver(gCtx global.Context, ctx context.Context, emote *struct
 
 	return &EmoteResolver{
 		ctx:          ctx,
+		quota:        ctx.Value(helpers.QuotaKey).(*helpers.Quota),
```
Comment (TroyDota) [inline src/server/v3/resolvers/query/emote.go:37]: rather reference a `utils.Key` value.

**Review by TroyDota — CHANGES_REQUESTED** [review]

File: src/server/v3/gql-v3.go [hunk 10]
```diff
@@ -98,6 +123,35 @@ func GQL(gCtx global.Context, app fiber.Router) {
...
+			})
+			b, _ := json.Marshal(usage)
+
+			c.Set("X-Quota-Usage", utils.B2S(b))
+		}
+		if !quota.Check() {
+			// Temporarily block this query
+			pipeline := gCtx.Inst().Redis.RawClient().Pipeline()
+			pipeline.SetEX(ctx, redisKey, "", time.Hour)
+			pipeline.HSet(ctx, fmt.Sprintf("%s:blocked-queries", instance.RedisPrefix), qh, clientIP)
```
Comment (TroyDota) [inline src/server/v3/gql-v3.go:144]: why are we setting this value in the map?
This will never expire correct?

Comment (AnatoleAM) [reply src/server/v3/gql-v3.go]: Is this solved by c135a06?

**Pull request merged**